src/query/storage/prom_converter.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"sync"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/cost"
+	"github.com/m3db/m3/src/query/generated/proto/prompb"
+	"github.com/m3db/m3/src/query/models"
+	xcost "github.com/m3db/m3/src/x/cost"
+	"github.com/m3db/m3/src/x/ident"
+	xsync "github.com/m3db/m3/src/x/sync"
+)
+
+func tagIteratorToLabels(
+	identTags ident.TagIterator,
+) ([]prompb.Label, error) {
+	labels := make([]prompb.Label, 0, identTags.Remaining())
+	for identTags.Next() {
+		identTag := identTags.Current()
+		labels = append(labels, prompb.Label{
+			Name:  identTag.Name.Bytes(),
+			Value: identTag.Value.Bytes(),
+		})
+	}
+
+	if err := identTags.Err(); err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+func iteratorToPromResult(
+	iter encoding.SeriesIterator,
+	enforcer cost.ChainedEnforcer,
+	tagOptions models.TagOptions,
+) (*prompb.TimeSeries, error) {
+	labels, err := tagIteratorToLabels(iter.Tags())
+	if err != nil {
+		return nil, err
+	}
+
+	samples := make([]prompb.Sample, 0, initRawFetchAllocSize)
```

Comment:
Because we can't really predict the sizes of the slice we need here, we can run into problems where a certain query needs to create huge slices, which get released to the pool but never shrunk back. We could potentially use a pool that has a chance to release and re-alloc smaller slices periodically, but not super sure how